From 84a1680004dc6559a58c314458c95d547e19df5e Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Tue, 4 Nov 2014 20:42:16 +0800
Subject: [PATCH 0145/5242] MLK-9794-3 usb: common: otg: HNP polling condition
 update for b host

commit  c91782c1404831ab93030311e6396afd7423e17b from
https://source.codeaurora.org/external/imx/linux-imx.git

If it's A host and b_hnp_enable is set, or B host, we can do HNP polling
because B host will not check OTG descriptor and set b_hnp_enable flag
when enumeration.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 6148c9df6d1c68dfc91cfe274101058c8edf979b)
(cherry picked from commit 8dbe217d14726d5950e91de9d98e401100357a21)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/common/usb-otg-fsm.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/common/usb-otg-fsm.c b/drivers/usb/common/usb-otg-fsm.c
index 8eabe5a..88c29a9 100644
--- a/drivers/usb/common/usb-otg-fsm.c
+++ b/drivers/usb/common/usb-otg-fsm.c
@@ -133,7 +133,8 @@ static void otg_hnp_polling_work(struct work_struct *work)
 	int retval;
 	struct usb_otg_descriptor *desc = NULL;
 
-	if (state != OTG_STATE_A_HOST && state != OTG_STATE_B_HOST)
+	if ((state != OTG_STATE_A_HOST || !fsm->b_hnp_enable) &&
+	    state != OTG_STATE_B_HOST)
 		return;
 
 	udev = usb_hub_find_child(fsm->otg->host->root_hub, 1);
-- 
1.7.9.5

